### ISSUE TITLE
[BUG]In the keydb flash mode, if the maxmemory is full, it will cause new data to be unable to write.

### ISSUE THREAD

**Issue by michael19960921** [Snapchat/KeyDB#652]
Hi, I've deployed a 6-node cluster with 3 master nodes and 3 slave nodes. Here's part of my configuration:
```
maxmemory 10mb
maxmemory-policy allkeys-lru
storage-provider flash /opt/keydb/flash/16380/
flash-disable-key-cache yes
storage-provider-options use_direct_reads=true;allow_mmap_reads=false;use_direct_writes=true;allow_mmap_writes=false
```
I enabled flash storage and set the maximum memory limit to 10MB, just to test the write performance of flash storage when it's full. However, during testing, I found that when I continuously write data until the memory is full, I get OOM exceptions and can't write any more new data. I suspect that the maxmemory-policy is not working and data is not being evicted.
My expectation is that when the memory is full, data continues to be stored on the disk and the data in memory is evicted. But this is not happening right now.
So, I disabled the flash mode and tested the memory scenario. With the memory still set to 10MB, I found that in this case, I could continuously write data, and the maxmemory-policy started working again.
Based on this, I suspect that there might be a problem with the maxmemory-policy in flash mode, and it's not working correctly. Could you please help?

**Comment by michael19960921** [comment]
Could someone please help me with my question? Thanks.

**Comment by msotheeswaran** [comment]
This is a bug, see #646.

**Comment by michael19960921** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by michael19960921** [comment]
I observed your submission records and compiled the latest main branch, but after I tried it, the issue still persists; when the memory is full, it is unable to write data and an OOM error occurs.

**Comment by michael19960921** [comment]
@msotheeswaran hi，I observed your submission records and compiled the latest main branch, but after I tried it, the issue still persists; when the memory is full, it is unable to write data and an OOM error occurs.

**Comment by msotheeswaran** [comment]
@michael19960921 are you using expires, the expireset could also cause the same issue.

**Comment by michael19960921** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by michael19960921** [comment]
@msotheeswaran I believe I understand your meaning, yes, I have set an expiration time for each key. So, you are implying that the expiration time setting could lead to an OOM? Then, this expiration time setting causing an OOM issue would be considered a bug. I would like to ask, which of the pull requests currently in progress is for fixing the expiration time? Flash expires?

**Comment by msotheeswaran** [comment]
@michael19960921 there is an in progress fix, but you can use these commits for now: https://github.com/Snapchat/KeyDB/commit/6eb595d6b4295f7a154367c8187d5c5d8eefef30 https://github.com/Snapchat/KeyDB/commit/6a32023cfc11bd7ce730b7486acec5f26bdeda49

**Comment by michael19960921** [comment]
@msotheeswaran 
After attempting to recompile and test the content submitted with 6eb595d 6a32023, I found that compared to the same memory (3 master nodes, 3 slave nodes, maximum memory 10MB) I was able to write more data, approximately 80%, however, I still encountered an Out Of Memory (OOM) error, and was unable to write any more data. Therefore, there may still be an issue preventing data from being written.

>  info memory
# Memory
used_memory:13536656
used_memory_human:12.91M
used_memory_rss:23232512
used_memory_rss_human:22.16M
used_memory_peak:13665384
used_memory_peak_human:13.03M
used_memory_peak_perc:99.06%
used_memory_overhead:8525840
used_memory_startup:4846864
used_memory_dataset:5010816
used_memory_dataset_perc:57.66%
allocator_allocated:16952184
allocator_active:19054592
allocator_resident:23035904
total_system_memory:8201158656
total_system_memory_human:7.64G
used_memory_lua:37888
used_memory_lua_human:37.00K
used_memory_scripts:0
used_memory_scripts_human:0B
number_of_cached_scripts:0
maxmemory:10485760
maxmemory_human:10.00M
maxmemory_policy:allkeys-lru
allocator_frag_ratio:1.12
allocator_frag_bytes:2102408
allocator_rss_ratio:1.21
allocator_rss_bytes:3981312
rss_overhead_ratio:1.01
rss_overhead_bytes:196608
mem_fragmentation_ratio:1.72
mem_fragmentation_bytes:9695904
mem_not_counted_for_evict:0
mem_replication_backlog:1048576
mem_clients_slaves:20512
mem_clients_normal:61536
mem_aof_buffer:0
mem_allocator:jemalloc-5.2.1
active_defrag_running:0
lazyfree_pending_objects:0
lazyfreed_objects:0
storage_provider:flash
flash_memory:0

**Comment by michael19960921** [comment]
@msotheeswaran hi, do you know the cause? My issue remains unresolved. I hope you have the time to answer me, thank you.

**Comment by msotheeswaran** [comment]
hi @michael19960921 I have not found the cause yet, I will update here when I have found/fixed the issue.

**Comment by michael19960921** [comment]
hi, @msotheeswaran Much appreciated for your hard work. Thank you very much for your attention to this matter.

**Comment by michael19960921** [comment]
hi， @msotheeswaran Has this bug been resolved?

**Comment by michael19960921** [comment]
hi， @msotheeswaran Has this bug been resolved?

**Comment by michael19960921** [comment]
hi， @msotheeswaran-sc Has this bug been resolved?

**Comment by msotheeswaran-sc** [comment]
Hi @michael19960921 all known memory bottlenecks for flash mode have been moved off memory, so this should be resolved with the next version release(6.3.4).

**Comment by msotheeswaran-sc** [comment]
Hi @michael19960921 I was hoping to get it out by the end of the month, but have been busy with internal work, will hopefully be out by mid november.
